### DISCUSSION TITLE
Upserting with with optional fields

### DISCUSSION THREAD

**Discussion by pandemonia** [mikro-orm/mikro-orm#5743] [Q&A] [answered]
I am trying to **upsert** with optional values excluded from the data, and hitting an error because the insert part of the `insert on conflict merge` errors out due to all values not being provided. 

Here's a minimal entity that can serve as an example:
```ts
@Entity()
class User {
  @PrimaryKey()
  id!: number;
 
  @Property()
  name!: string;

  @Property()
  bar: Date & Opt = new Date();
}
```

With `bar` having a default value, I can **create** an entity with just `em.create(User, {id: 1, name: 'test'})`, omitting `bar` here. This gives us the query

```sql
insert into `user` (`id`, `name`, `bar`) values (1, 'test', 1704067200000)
```

with the ORM supplying the `bar` value to the DB.

Continuing the same train of thought with **upsert**, attempting `em.upsert(User, {id: 1, name: 'test'})` gives us this:

```sql
insert into `user` (`id`, `name`) values (1, 'test') on conflict (`id`) do update set `name` = excluded.`name` returning `bar`
```

which fails with `NOT NULL constraint failed: user.bar`. 

Looking further at the query, this **upsert** query should probably result in

```sql
insert into `user` (`id`, `name`, `bar`) values (1, 'test', 1704067200000) on conflict (`id`) do update set `name` = excluded.`name` returning `bar`
```

with the `bar` field's default value supplied to the insert clause similar to the insert query, but excluded in the `do update`.

While I can achieve this query manually by setting the upsertOptions, this seems like something the ORM should be setting on its own - the data previously passed in the **upsert** has all the required fields; and this object working with **create** but failing in **upsert** seems inconsistent.

Manual workaround that I can see:
```ts
em.upsert(
  User,
  {id: 1, name: 'test', bar: new Date()},
  {onConflictExcludeFields: ['bar]}
);
```

I have also given some thought on how this could be extended to **upsertMany** but am drawing a blank there!

We have multiple workflows periodically ingesting a large number of rows, and having the upsert/upsertMany would be the most efficient implementation. I have a full reproduction here for reference: https://github.com/pandemonia/upsert-default-fields.

Thanks for this library - my first month with this and its been serving us well so far.

**Comment by B4nan** [comment]
> Looking further at the query, this upsert query should probably result in

Runtime defaults do not have any effect on `em.upsert(Type, data)`, since there are no entity instances created before the query is fired. But you can create the entity yourself, then its property initializers will get executed, and you can do `em.upsert(entity)` (or `em.upsertMany([e1, e2, ...])`) instead.

**Reply by pandemonia** [reply]
> since there are no entity instances created before the query is fired.

That makes sense.

>  But you can create the entity yourself, then its property initializers will get executed, and you can do `em.upsert(entity)`

This would do the insert correctly, but updating with a created entity with all values initialised will override the optional fields with the newly set default values. What I was hoping `em.upsert(User, { id, name })` would be was:

### Create if no existing entity with id
```sql
insert into `user` (`id`, `name`, `bar`) values (1, `test`, 1704067200000); 
```

_or_

### Update existing entity
```sql
update `user` set `name` = `test` where `id` = 1;
```
without overriding the `bar` value for the existing entity.

**Reply by B4nan** [reply, marked as answer]
What about using db default on the property? 

```ts
@Property({ default: sql.now() })
bar: Date & Opt = new Date();
```

Then you should be able to use `em.upsert(User, { id, name })` the way you want, since its fine to not provide value for columns with defaults.

**Reply by pandemonia** [reply]
Yes, this works. Thanks for the help!
